### PR TITLE
Make dc/rack encryption work for some cases where Nat hides ednpoint ips

### LINKED ISSUE #9653: Node fails to start: Operation timed out for system_distributed_everywhere.cdc_generation_descriptions_v2
*Installation details*
Scylla version (or git commit hash): 4.6.dev-0.20211107.4950ce539 with build-id 8d7a7f85964575a8775f8358cef3ff74141e07fd
Cluster size: 9, split between 3 dc's: 4, 3 and 2 on eu-west-1, us-west-2, us-east-1 respectively.
OS (RHEL/CentOS/Ubuntu/AWS AMI): ami-059ce9f5311c1f755 ami-0515872fec42d763c ami-09170b5e7a8dbaea9 on eu-west-1, us-west-2, us-east-1 respectively.
Test case: https://github.com/scylladb/scylla-cluster-tests/blob/master/test-cases/longevity/longevity-lwt-24h-multidc.yaml

The first node of the run (54.72.180.96 | 10.0.1.198) started with no issue. The second node (54.75.114.234 | 10.0.1.39), however, failed to start altogether, due to the following:
```
2021-11-15T12:05:55+00:00 lwt-longevity-multi-dc-24h-4-5-db-node-d67a2920-2 !ERR | scylla: [shard 0] init - Startup failed: exceptions::mutation_write_timeout_exception (Operation timed out for system_distributed_everywhere.cdc_generation_descriptions_v2 - received only 1 responses from 2 CL=ALL.)
```

The nodes were kept alive:
```
lwt-longevity-multi-dc-24h-4-5-db-node-d67a2920-1 [54.72.180.96 | 10.0.1.198]
lwt-longevity-multi-dc-24h-4-5-db-node-d67a2920-2 [54.75.114.234 | 10.0.1.39]
lwt-longevity-multi-dc-24h-4-5-db-node-d67a2920-3 [34.241.165.213 | 10.0.2.233]
lwt-longevity-multi-dc-24h-4-5-db-node-d67a2920-4 [3.250.167.195 | 10.0.2.188] 
lwt-longevity-multi-dc-24h-4-5-db-node-d67a2920-5 [35.86.177.1 | 10.0.1.87] 
lwt-longevity-multi-dc-24h-4-5-db-node-d67a2920-6 [52.33.121.139 | 10.0.1.93] 
lwt-longevity-multi-dc-24h-4-5-db-node-d67a2920-7 [54.184.35.197 | 10.0.0.139] 
lwt-longevity-multi-dc-24h-4-5-db-node-d67a2920-8 [3.239.70.88 | 10.0.3.146] 
lwt-longevity-multi-dc-24h-4-5-db-node-d67a2920-9 [3.239.222.71 | 10.0.1.171]
```

### Please let me know when I can terminate them

Logs:
```
+-----------------------------------------------------------------------------------------------------------------------------------------------------------------+
|                                             Log links for testrun with test id d67a2920-93bc-4fd1-9c2d-3b9d60b54f98                                             |
+-----------------+-------------+---------------------------------------------------------------------------------------------------------------------------------+
| Date            | Log type    | Link                                                                                                                            |
+-----------------+-------------+---------------------------------------------------------------------------------------------------------------------------------+
| 20211115_130630 | db-cluster  | https://cloudius-jenkins-test.s3.amazonaws.com/d67a2920-93bc-4fd1-9c2d-3b9d60b54f98/20211115_130630/db-cluster-d67a2920.tar.gz  |
| 20211115_130630 | loader-set  | https://cloudius-jenkins-test.s3.amazonaws.com/d67a2920-93bc-4fd1-9c2d-3b9d60b54f98/20211115_130630/loader-set-d67a2920.tar.gz  |
| 20211115_130630 | monitor-set | https://cloudius-jenkins-test.s3.amazonaws.com/d67a2920-93bc-4fd1-9c2d-3b9d60b54f98/20211115_130630/monitor-set-d67a2920.tar.gz |
| 20211115_130630 | sct-runner  | https://cloudius-jenkins-test.s3.amazonaws.com/d67a2920-93bc-4fd1-9c2d-3b9d60b54f98/20211115_130630/sct-runner-d67a2920.tar.gz  |
+-----------------+-------------+---------------------------------------------------------------------------------------------------------------------------------+
```

### PR DESCRIPTION
**Pull request by eliransin**
This is a consolidation of #9714 and #9709 PRs by @elcallio that were reviewed by @asias
The last comment on those was that they should be consolidated in order not to create a security degradation for
ec2 setups.
For some cases it is impossible to determine dc or rack association for nodes on outgoing connections.
One example is when some IPs are hidden behind Nat layer.
In some cases this creates problems where one side of the connection is aware of the rack/dc association where the
other doesn't.
The solution here is a two stage one:
1. First add a gossip reverse lookup that will help us determine the rack/dc association for a broader (hopefully all) range
 of setups and NAT situations.
2. When this fails - be more strict about downgrading a node which tries to ensure that both sides of the connection will at least
 downgrade the connection instead of just fail to start when it is not possible for one side to determine rack/dc association.

Fixes #9653
/cc @elcallio @asias

### CODE REVIEW

**Comment by eliransin** [comment]
The two PRs reference here were consolidated as a resolvance for @asias final review comment, @asias reviewed those PRs
to @avikivity request.

**Comment by avikivity** [comment]
I don't see the reverse mapping patch here.

**Comment by avikivity** [comment]
(the two pull requests had 3 patches in total, this has 2).

**Comment by eliransin** [comment]
> (the two pull requests had 3 patches in total, this has 2).

Will add it now. I also saw your comments on the individual PRs.

File: message/messaging_service.hh [hunk 1]
```diff
@@ -274,7 +274,7 @@ private:
 private:
     config _cfg;
     // map: Node broadcast address -> Node internal IP for communication within the same data center
-    std::unordered_map<gms::inet_address, gms::inet_address> _preferred_ip_cache;
+    std::unordered_map<gms::inet_address, gms::inet_address> _preferred_ip_cache, _preferred_to_endpoint;
```
Comment (asias) [inline message/messaging_service.hh:277]: The comments is wrong for _preferred_to_endpoint. Let's put it on a new line and add correct comments.

Comment (asias) [reply message/messaging_service.hh:277]: Original discussion is here here: https://github.com/scylladb/scylla/pull/9714#discussion_r764072237

Comment (eliransin) [reply message/messaging_service.hh:277]: I have changed the comment to indicate that the second member is the reversed mapping.

**Comment by eliransin** [comment]
@avikivity all review comments were addressed. The only thing left is to see if you are satisfied with the answers to the
concern you raise in https://github.com/scylladb/scylla/pull/9714#pullrequestreview-835873291
If you are, I am not aware of anything preventing this from being merged.

**Comment by slivne** [comment]
@avikivity ping ^ 2

**Pull request closed without merging**